effect/test_async.py

```diff
@@ -0,0 +1,84 @@
+from __future__ import print_function
+
+from testtools.matchers import raises
+from testtools.testcase import TestCase
+
+from characteristic import attributes
+
+from ._base import Effect, perform
+from ._dispatcher import ComposedDispatcher, TypeDispatcher
+from ._intents import (
+    Constant, Error, FirstError, ParallelEffects, parallel, base_dispatcher)
+from ._sync import sync_perform
+from .async import perform_parallel_async
+from .test_base import func_dispatcher
+
+@attributes(['message'])
+class EquitableException(Exception):
+    pass
+
+
+
+class PerformParallelAsyncTests(TestCase):
+
+    def setUp(self):
+        super(PerformParallelAsyncTests, self).setUp()
+        self.dispatcher = ComposedDispatcher([
+            base_dispatcher,
+            TypeDispatcher({ParallelEffects: perform_parallel_async})])
+
+    def test_empty(self):
+        """
+        When given an empty list of effects, ``perform_parallel_async`` returns
+        an empty list synchronusly.
+        """
+        result = sync_perform(
+            self.dispatcher,
+            parallel([]))
+        self.assertEqual(result, [])
+
+    def test_parallel(self):
+        """
+        'parallel' results in a list of results of the given effects, in the
+        same order that they were passed to parallel.
+        """
+        result = sync_perform(
+            self.dispatcher,
+            parallel([Effect(Constant('a')),
+                      Effect(Constant('b'))]))
+        self.assertEqual(result, ['a', 'b'])
+
+    def test_error(self):
+        """
+        When given an effect that results in a Error,
+        ``perform_parallel_async`` result in ``FirstError``.
+        """
+        self.assertThat(
+            lambda: sync_perform(
+                self.dispatcher,
+                parallel([Effect(Error(EquitableException(message="foo")))])),
+            raises(FirstError(exception=EquitableException(message='foo'),
```

Comment:
reposting this comment since it got eaten by github:

this is a bogus assertion. it's not checking equality with EquitableException. Also, need to test a non-0 index.